src/mix.cr

```diff
@@ -0,0 +1,255 @@
+require "./lib_mix"
+require "./sdl"
+
+module SDL
+  module MIX
+    alias Init = LibMIX::Init
+    MAX_VOLUME = LibMIX::MIN_MAX_VOLUME
+    
+    enum Type
+      AIFF
+      FLAC
+      MIDI
+      MOD
+      MP3
+      OGG
+      VOC
+      WAV
+    end
+
+    # Used to load the support for the flags. `quit` must be called during app cleanup
+    def self.init(flags : Init)
+      ret = LibMIX.init(flags)
+      unless (ret & flags.value) == flags.value
+        raise SDL::Error.new("Mix_Init failed to init #{flags}")
+      end
+    end
+
+    def self.quit
+      LibMIX.quit
+    end
+
+    # This is required to initial SDL_Mixer. It must be called before using any other function, but AFTER SDL has been initialized.
+    def self.open(freq = 44100, format = 0_u16, channels = 2, sample_size = 2048)
+      format = format == 0 ? LibMIX::MIX_DEFAULT_FORMAT : format.to_u16
+      ret = LibMIX.open_audio(freq, format, channels, sample_size)
+      raise SDL::Error.new("Mix_OpenAudio") unless ret == 0
+      ret
+    end
+
+    def self.query_spec(freq = 44100, format = 0_u16, channels = 2)
+      audio_open_count = LibMIX.query_spec(freq, format, channels)
+      raise SDL::Error.new("Mix_QuerySpec: #{LibMIX.get_error}") if audio_open_count < 1
+      audio_open_count
+    end
+
+    def self.close
+      LibMIX.close_audio
+    end
+
+  end
+
+  module MIX
+    class Sample
+      protected getter sample
+
+      def self.decoder_count
+        LibMIX.get_num_chunk_decoders
+      end
+
+      def self.decoder_name(index)
+        LibMIX.get_chunk_decoder(index)
+      end
+
+      def initialize(filename = nil)
+        load filename if filename
+      end
+
+      def load(filename)
+        rwops = LibSDL.rw_from_file(filename, "rb")
+        @sample = LibMIX.load_wav_rw(rwops, 1)
+        raise SDL::Error.new("Mix_LoadWAV_RW") unless @sample
+      end
+
+      def finalize
+        LibMIX.free_chunk(sample)
+      end
+    end
+  end
+
+  module MIX
+    class Music
+      private getter music : Pointer(LibMIX::Music) | Nil
+
+      def initialize(filename = nil)
+        @music = filename ? load(filename) : nil
```

Comment:
Same here: `MIX::Music#initialize` should require a filename and always load a music file; `@music` doesn't have to be a nilable, and `#finalize` doesn't have to set it to nil (GC is collecting the memory anyway).

Unless there is something I didn't understand about SDL_Mixer... like we may only play a single BGM music at once, in which case we should have an argless `MIX::Music#initialize` (i.e. remove the method) have `#load` load or replace the current music (which may require to free any previous `@music`?) then verify that `@music` is set before calling `free_music` in `#finalize`.

I even wonder if it should be a class at all, or just a module with class methods? There is no state, except the currently loaded `SDL::RWops` and `SDL::MIX::Music` which could be class variables.